File: HW01/Cylinder_Series.cxx
// This is an example about to generate 10 sylinders in a row.
//
#include "vtkSmartPointer.h"
#include "vtkConeSource.h"
#include "vtkSphereSource.h"
#include "vtkPolyDataMapper.h"
#include "vtkActor.h"
#include "vtkRenderer.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkProperty.h"
#include "vtkCamera.h"
#include "vtkLight.h"
#include "unistd.h"

int main()
{
  
  

  vtkConeSource *sphere = vtkConeSource::New();
  sphere->SetHeight( 3.0 );
  sphere->SetRadius( 1.0 );
  sphere->SetResolution( 10 );
  vtkSmartPointer<vtkPolyDataMapper> sphereMapper = vtkSmartPointer<vtkPolyDataMapper>::New();
  sphereMapper->SetInputConnection(sphere->GetOutputPort());
  vtkSmartPointer<vtkRenderWindowInteractor> iren = vtkSmartPointer<vtkRenderWindowInteractor>::New();
  vtkSmartPointer<vtkRenderWindow> renWin = vtkSmartPointer<vtkRenderWindow>::New();
  renWin->SetSize(400, 200);
  iren->SetRenderWindow(renWin);
  vtkSmartPointer<vtkLight> light = vtkSmartPointer<vtkLight>::New();
  light->SetFocalPoint(1.875, 0.6125, 0);
  light->SetPosition(0.875, 1.6125, 1);

  

  
  

  // Wait for 100 ms to show next group
  for (unsigned int j = 0; j < 360; j++)
    {
      //sleep(0.03);
      // for iteration to get 4 spheres.
      //
      vtkSmartPointer<vtkRenderer> ren = vtkSmartPointer<vtkRenderer>::New();
      renWin->AddRenderer(ren);
      for(unsigned int i = 0; i < 4; i++)
	{
	  vtkSmartPointer<vtkActor> sphere1 = vtkSmartPointer<vtkActor>::New();
	  sphere1->SetMapper(sphereMapper);
	  sphere1->GetProperty()->SetColor(1, 0.1*i,0);
	  sphere1->GetProperty()->SetAmbient(0.1*i);
	  sphere1->AddPosition(i*10, 0, 0);

	  // add actors to the renderer.
	  ren->AddActor(sphere1);
	}
      ren->SetBackground(0.1, 0.2, 0.4);
      ren->GetActiveCamera()->Azimuth(1);
      // set up the lighting.
      
      ren->AddLight(light);

    }
  iren->Initialize();
  iren->Start();
  
 
  

  return EXIT_SUCCESS;
}

File: HW01/Cylinder_Series.py
"""
python(2.7) version of Cylinder_series.

Zhewei @ 9/25/2016

"""
from __future__ import print_function
import vtk
import sys, gzip
import pickle as Pickle
import numpy

bottomLine1 = -100
bottomLine2 = 30
distance_eachSubjects = 20
center_of_Subjects = 200
augment_ratio = 100
timeFrame = 130
featureNo = 20



def source_to_mapper_to_actor(height):
    source = vtk.vtkCylinderSource()
    mapper = vtk.vtkPolyDataMapper()
    actor = vtk.vtkActor()
    source.SetCenter(0,0,0)
    source.SetRadius(6.0)
    source.SetHeight(height*augment_ratio)
    source.SetResolution(100)
    mapper.SetInputConnection(source.GetOutputPort())
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(0,1,0)
    return source, mapper, actor

def main(args):
    if len(args) < 3:
        usage( args[0] )
        pass
    else:
        work( args[1:] )
        pass
    pass

def usage (programm):
    print ("usage: python2.7 %s numer1 number2"%(programm))
    print ('       number1 and number2 are integers in [0,180]')

def work(numbers):
    Raw_data = gzip.open('VTK_Subjects_180.pickle.gz', 'rb')
    Subjects_data, Label, ID = Pickle.load(Raw_data)
    subjectNo1 = int(numbers[0])
    subjectNo2 = int(numbers[1])
    print ("First subject is", Label[subjectNo1])
    print ('Second subject is', Label[subjectNo2])

    data1 = Subjects_data[subjectNo1*130:(subjectNo1+1)*130,:]
    data2 = Subjects_data[subjectNo2*130:(subjectNo2+1)*130,:]
    # create a rendering window and renderer
    renWin = vtk.vtkRenderWindow()
    iren = vtk.vtkRenderWindowInteractor()
    camera = vtk.vtkCamera()
    camera.SetPosition(0, (bottomLine1+bottomLine2)/2, -550)


    for j in range(1,timeFrame-1):
        ren = vtk.vtkRenderer()
        ren.SetActiveCamera(camera)
        for i in range(featureNo):
            renWin.AddRenderer(ren)
 
            # create source
            height1 = data1[j,i]
            height2 = data2[j,i]
            source1, mapper1, actor1 = source_to_mapper_to_actor(height1)
            source2, mapper2, actor2 = source_to_mapper_to_actor(height2)
                 
            actor1.AddPosition(distance_eachSubjects*i-center_of_Subjects, height1*100/2+bottomLine1,0)
            actor2.AddPosition(distance_eachSubjects*i-center_of_Subjects, height2*100/2+bottomLine2,0)

            # assign actor to the renderer
            ren.AddActor(actor1)
            ren.AddActor(actor2)
            ren.SetBackground(0.1, 0.2, 0.4)
        renWin.AddRenderer(ren)
        renWin.SetSize(700,400)

        renWin.Render()


        
if __name__ == '__main__':
    main(sys.argv)
    pass
